doc: update technical escalation proccess (#63587)
zephyrproject-rtos/zephyr: doc/contribute/contributor_expectations.rst

defines the following escalation process for resolving technical disagreements.

Before escalation of technical disagreements, follow the steps below:

- Resolve in the PR among assignee, maintainers and reviewer.

  - Assignee to act as moderator if applicable.

- Optionally resolve in the next `Zephyr Dev Meeting`_  meeting with more
  Maintainers and project stakeholders.

  - The involved parties and the Assignee to be present when the  issue is
    discussed.

- If no progress is made, the assignee (maintainer) has the right to dismiss
  stale, unrelated or irrelevant change requests by reviewers giving the
  reviewers a minimum of 1 business day to respond and revisit their initial
  change requests or start the escalation process.

  The assignee has the responsibility to document the reasoning for dismissing
  any reviews in the PR and should notify the reviewer about their review being
  dismissed.

  To give the reviewers time to respond and escalation, the assignee is
  expected to block the PR from being merged either by not
  approving the PR or by setting the *DNM* label.

Escalation can be triggered either by the assignee or the reviewers requesting

> changes following the steps below:

Reviewer: Escalation by a the person who initiated the PR should also be allowed.

> - Escalate to the `Architecture Working Group`_ by adding the `Architecture

Reviewer 2: Can we make this architecture working discussion optional together with the Zephyr Dev Meeting? so we don't have both the arch working group and the TSC?

Author: I prefer to have this resolved much earlier and not having to deal with PR conflicrt resolution in the TSC. The TSC is the  very very last resort and should not be seen as the first thing you can do. I also do not see the Dev Meeting on the path for escalation, it is a discussion forum and it will be less confusing if we had less choices and less "or"s IMO.

In cases where a contributor objects to change requests from reviewers, Zephyr
defines the following escalation process for resolving technical disagreements.

Before escalation of technical disagreements, follow the steps below:

- Resolve in the PR among assignee, maintainers and reviewer.

  - Assignee to act as moderator if applicable.

- Optionally resolve in the next `Zephyr Dev Meeting`_  meeting with more
  Maintainers and project stakeholders.

  - The involved parties and the Assignee to be present when the  issue is
    discussed.

- If no progress is made, the assignee (maintainer) has the right to dismiss
  stale, unrelated or irrelevant change requests by reviewers giving the
  reviewers a minimum of 1 business day to respond and revisit their initial
  change requests or start the escalation process.

  The assignee has the responsibility to document the reasoning for dismissing
  any reviews in the PR and should notify the reviewer about their review being
  dismissed.

>   To give the reviewers time to respond and escalation, the assignee is

Reviewer 3: s/escalation/escalate/

  expected to block the PR from being merged either by not

>   approving the PR or by setting the *DNM* label.

Reviewer 4: There isn't any enforcement that assignees have approved a PR, correct?  I would simplify this language to just apply the DNM label.

Escalation can be triggered by any party participating in the review
process (assignee, reviewers or the original author of the change)  following
the steps below:

- Escalate to the `Architecture Working Group`_ by adding the `Architecture
  Review` label on the PR. Beside the weekly meeting where such escalations are
  processed, the `Architecture Working Group`_  shall facilitate an offline
  review of the escalation if requested, especially if any of the parties can't
  attend the meeting.

> - If issues are not resolved, assignees can escalate to the TSC and get a
>   binding resolution in the TSC by adding the *TSC* label on the PR.

Reviewer 4: You mention in a comment that the TSC is a last resort.  How about making that explicit here? 
```suggestion
- If all avenues of resolution and escalation have failed, assignees can escalate 
  to the TSC and get a binding resolution in the TSC by adding the *TSC* label
  on the PR.
```

  reviewers a minimum of 1 business day to respond and revisit their initial
  change requests or start the escalation process.

  The assignee has the responsibility to document the reasoning for dismissing
  any reviews in the PR and should notify the reviewer about their review being
  dismissed.

  To give the reviewers time to respond and escalation, the assignee is
  expected to block the PR from being merged either by not
  approving the PR or by setting the *DNM* label.

Escalation can be triggered by any party participating in the review
process (assignee, reviewers or the original author of the change) following
the steps below:

- Escalate to the `Architecture Working Group`_ by adding the `Architecture
  Review` label on the PR. Beside the weekly meeting where such escalations are
  processed, the `Architecture Working Group`_  shall facilitate an offline
  review of the escalation if requested, especially if any of the parties can't
  attend the meeting.

- If all avenues of resolution and escalation have failed, assignees can escalate
  to the TSC and get a binding resolution in the TSC by adding the *TSC* label

>   on the PR.

Reviewer 5: Far from being polemic, I believe we can solve all the disagreements earlier in the regular AG meeting. However, sometimes it can happen, and TSC's decision is vital. Is there any process regarding that? Will the TSC decide based on what information? Should the involved parties invoked to tell their point of view or something similar? 

Based on my situation as an independent maintainer. I won't have any direct representatives (as a company or association) on the TSC members. For me, It would be harder to succeed in a disagreement with a regular employee of a Zephyr platinum-member company, for example. Is that the intention, or are we trying to be more democratic?

Reviewer 6: There are two community members in TSC.

Reviewer 5: > There are two community members in TSC.

Thanks. Great to know. I tried to see who are they, but I could not.

Reviewer: Me and @erwango.

Reviewer 6: There is even c in TSC for community, and small _n_ somewhere for neutrality.

  - Assignee to act as moderator if applicable.

> - Resolve in the next `Zephyr Dev Meeting`_  meeting with more

Reviewer 2: Please leave the "optionally" (and ideally the whole paragraph as it was). As you @nashif said not everybody can join these meetings. We don't all live in US time. It is hardly helpful to demand that 2 people in Europe, India or Australia need to join a meeting in US pacific time in which things may or may not be discussed, in which there is high likelihood nobody else will care about the issue.
Having the dev-review and architecture meeting as an option/possibility for discussion is good. Forcing it as the path, no.

  Maintainers and project stakeholders.

  - The involved parties and the Assignee to be present when the  issue is
    discussed.

- If the above steps do not yield any results, the assignee (maintainer) has
  the right to dismiss stale, unrelated or irrelevant change requests by
  reviewers giving the reviewers a minimum of 1 business day to respond and

>   revisit their initial change requests or escalate to the `Architecture

Reviewer 2: Why the tripple escalation of dev-review, architecture working group and then TSC? (This seems like a lot of steps to keep dragging matters forever).
I would leave this as it was before (i.e. optionally discuss it in the dev-review or arch workgroup, then TSC).

>   Working Group`_.

Reviewer 7: The minimum of 1 business day to respond is by responding in the PR I believe but not sure how to tie this with the "or" escalate to the arch WG? Should this be, 1 day to respond and if objecting, escalate to the next arch WG meeting?

Reviewer 8: Yeah I also find the dynamics of this a bit awkward, how does merging plays into this? Review dismissed, should merging wait some time for the person to escalate?

Author: addressed....

  - The involved parties and the Assignee to be present when
    the (escalated) issue is discussed.

- The assignee (maintainer) has the right to dismiss stale, unrelated or
  irrelevant change requests by reviewers giving the reviewers a minimum of 1
  business day to respond and revisit their initial change requests or escalate

>   to the architecture working group.

Reviewer 2: It was agreed long ago and approved in the TSC, that this was the case. That is, that a maintainer in an area can dismiss reviews from collaborators or people less involved in their area. And that in that case, those who disagree have the right to escalate to the TSC if they feel the need.
The idea is that maintainers are trusted to make decisions in their area to a large degree but an escalation path exists.
The situation otherwise was one in which anybody could block progress in an area even if they were not involved in it, and when a disagreement started, there was too much arm waving and other not proper behaviors that you'd expect from politicians and not engineers, which was not good at all. We had had some pointless confrontations before this rule, which this solved.
It is very rare that reviews are dismissed due to disagreements, and even rarer that they are escalated to the TSC.

Disenpowering maintainers and forcing any disagreement to be brought to a meeting in which most have no understanding of the issue would bring us back to a bad situation, so I would very much object to that. Those not involved in a review are typically those who have no stake or knowledge, so they should not be the default decision maker.

>   to the architecture working group.

Author: > I think cases that result in review dismissals should be carried on by someone unrelated to the author and reviewers.

this is exactly what the escalation process is about.

>   to the architecture working group.

Reviewer 8: I like the idea that maintainer are trusted and empowered to make decisions, but I think that there's few corner cases that should be considered, for instance:
- Driver changes usually falls between two (or more) maintainer areas: the subsystem and the platform, who end up assigned is pretty arbitrary, IIRC it's based on the file count but patform has priority. I think in case of disagreement the subsystem maintainer should have priority over the platform one.
- Situations where the assignee author and in the same org, the reviewer is external. It may be alright to dismiss but in principle it does not really play well with the vendor neutrality nature of the project.

>   to the architecture working group.

Reviewer 8: > this is exactly what the escalation process is about.

Oh you mean up to arch or tsc? Sure but then the dismissal could be up to the working group.

>   to the architecture working group.

Author: > Oh you mean up to arch or tsc? Sure but then the dismissal could be up to the working group.

no, because you might dismiss some review that is blocking a PR for no good reason and the reviewer concerned might not even care or they might just accept it and move on, we do not want to make every disagreement go to a WG, the WG or TSC path only needed when disagreement continues and no way to resolve in the PR.

>   to the architecture working group.

Author: > * Driver changes usually falls between two (or more) maintainer areas: the subsystem and the platform, who end up assigned is pretty arbitrary, IIRC it's based on the file count but patform has priority. I think in case of disagreement the subsystem maintainer should have priority over the platform one.

we already document that maintainer have the responsibility to re-assign. Anyone can contest wrong assignment and re-assign if it makes sense. But we also have to expect that 'maintainer' should be looking for the project and the common good by default and that maintainers will do the right thing at the end. There is no system that is perfect and we need to have some trust. 

> Situations where the assignee author and in the same org, the reviewer is external. It may be alright to dismiss but in principle it does not really play well with the vendor neutrality nature of the project.

This is the reality of the project and we have recently added the measure to at least not allow merges of PRs when submitter, reviewers are from the same org, so we have at least one pair of eyes that is not involved directly. Any other reviews from other collaborators are still and will always be welcome, but the reality here is that the vendors know their hardware best. As long as the context of a change is "local" to one vendor, wrong doing will only harm the vendor themselves. 
This become more critical for "common" areas, and we already request that more eyes from different organisation are part of the review.

>   to the architecture working group.

Reviewer 8: > [...] but the reality here is that the vendors know their hardware best. As long as the context of a change is "local" to one vendor, wrong doing will only harm the vendor themselves.

That sounds like a situation where having external contributions would improve the overall result. A feedback from someone unfamiliar with the specific hardware may highlight documentation gaps that would otherwise be missed. Consciously accepting harmful changes just because it's withing a vendor set of drivers also does not strike me as ideal.

Anyway, I think I may be misreading this, is the intended sequence like:
- assignee dismisses the review
- the reviewer has 1 day to follow up

or
- the assignee asks the reviewer to clarify or escalate to arch
- the review gets dismissed if no response after 1 day

?

>   to the architecture working group.

Reviewer 9: > That sounds like a situation where having external contributions would improve the overall result.

Yes of course and sometimes it happens which is unbelievable but this is realistically the situation where there is the least money and least interest to perform "outsider" reviews.

> Consciously accepting harmful changes...

Strike "consciously".

>   to the architecture working group.

Author: > Anyway, I think I may be misreading this, is the intended sequence like:
> 
> * assignee dismisses the review
> * the reviewer has 1 day to follow up

this.

There are many other steps and things that can be done before a review is dismissed, i.e. discussion in the PR itself, discord, the dev-review meeting etc or even the arch. WG if the issue at hand is at the level we want to have in this meeting, but ultimately and if all options have been considered and tried, the assignee can dismiss with a justification and eventually go through the escalation process if requested.

>   to the architecture working group.

Reviewer 8: > There are many other steps and things that can be done before a review is dismissed, i.e. discussion in the PR itself, discord, the dev-review meeting etc or even the arch. WG if the issue at hand is at the level we want to have in this meeting, but ultimately and if all options have been considered and tried, the assignee can dismiss with a justification and eventually go through the escalation process if requested.

Ok, I missed out that this was the last item in the escalation list and already after the "talk to each other" and "talk to each other in a working group" steps. Yeah I think this is fine then, though it's odd that the working group item is "optionally resolve...". I think a reviewer should  be given the chance to ask to bring up a PR in a working group before their review is dismissed. Maybe drop "optionally" from the previous step?

>   to the architecture working group.

Reviewer 7: Everything LGTM, but to @fabiobaltieri concerns, this section actually starts with resolving issues in the PR, then it says "optionally" take the conversation to the dev or arch meetings. Might I suggest that we remove the word "optionally" and just make the steps:

1. Work within the PR and community infrastructure (discord) to resolve the PR
2. Bring to the dev/arch meeting to discuss

Let the discussion results from the dev/arch be the trigger to decide if the review should be dismissed or not with the assignee being able to do the dismissal?

>   to the architecture working group.

Author: dev-review meeting is not part of the escalation, it is just a discussion forum and artch WG meeting should not be used to discuss every single disagreement in PRs, the backlog of technical discussion of new subsystems APIs etc is already long enough, this just does not scale, also, not everyone can attend those meetings. So escalation path should be short we defined and should not involve 3 meetings (dev-review, arch wg, TSC...).

Anyways, I added more updates, please take another look.

>   to the architecture working group.

Reviewer 8: Thanks I think it's a bit more clear now.

- If issues are not resolved, assignees can escalate to the TSC and get a

>   binding resolution in the TSC by adding the `tsc`_ label on the PR.

Reviewer: ```suggestion
  binding resolution in the TSC by adding the `TSC`_ label on the PR.
```

> - The Assignee is expected ensure the resolution of the escalation and the

Reviewer 2: typo: 
.. is expected +to+ ensure..
